src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp

```diff
@@ -72,6 +61,101 @@ bool SmokeCoAlarmServer::SetExpressedState(EndpointId endpointId, ExpressedState
     return success;
 }
 
+bool SmokeCoAlarmServer::AutoSetExpressedState(EndpointId endpointId,
+                                               std::array<ExpressedStateEnum, kPriorityOrderLength> priorityOrder)
+{
+    ExpressedStateEnum currentExpressedState = ExpressedStateEnum::kNormal;
+
+    for (ExpressedStateEnum priority : priorityOrder)
```

Comment:
These switches are very long, I wonder if they can somehow be shortened. Thinking something like:

```cpp
bool SmokeCoAlarmServer::GetStateActive(ExpressedStateEnum e, chip::EndpointId endpointId, bool & active)
{
    AlarmStateEnum state = AlarmStateEnum::kNormal;

    switch (e)
    {
    case ExpressedStateEnum::kSmokeAlarm:
        VerifyOrReturnValue(GetSmokeState(endpointId, state), false);
        active = state != AlarmStateEnum::kNormal;
        break;
    case ExpressedStateEnum::kCOAlarm:
        VerifyOrReturnValue(GetCOState(endpointId, state), false);
        active = state != AlarmStateEnum::kNormal;
        break;
    case ExpressedStateEnum::kBatteryAlert:
        VerifyOrReturnValue(GetBatteryAlert(endpointId, state), false);
        active = state != AlarmStateEnum::kNormal;
        break;
    case ExpressedStateEnum::kInterconnectSmoke:
        VerifyOrReturnValue(GetInterconnectSmokeAlarm(endpointId, state), false);
        active = state != AlarmStateEnum::kNormal;
        break;
    case ExpressedStateEnum::kEndOfService: {
        EndOfServiceEnum current = EndOfServiceEnum::kNormal;
        VerifyOrReturnValue(GetEndOfServiceAlert(endpointId, current), false);
        active = current != EndOfServiceEnum::kNormal;
        break;
    }
    case ExpressedStateEnum::kInterconnectCO:
        VerifyOrReturnValue(GetInterconnectCOAlarm(endpointId, state), false);
        active = state != AlarmStateEnum::kNormal;
        break;
    case ExpressedStateEnum::kTesting:
        VerifyOrReturnValue(GetTestInProgress(endpointId, active), false);
        break;
    case ExpressedStateEnum::kHardwareFault:
        VerifyOrReturnValue(GetHardwareFaultAlert(endpointId, active), false);
        break;
    default:
        active = false;
    }
    return true;
}

bool SmokeCoAlarmServer::AutoSetExpressedState(EndpointId endpointId,
                                               std::array<ExpressedStateEnum, kPriorityOrderLength> priorityOrder)
{
    ExpressedStateEnum currentExpressedState = ExpressedStateEnum::kNormal;

    for (ExpressedStateEnum priority : priorityOrder)
    {
        bool active = false;
        VerifyOrReturnValue(GetStateActive(priority, endpointId, active), false);
        if (active)
        {
            currentExpressedState = priority;
            break;
        }
    }

    return SetExpressedState(endpointId, currentExpressedState);
 }
 ```
 
 Whilch maybe expresses the intent clearer. The usage of bool status values still make the code feel somewhat odd, but at least this seems to explain the logic somewhat clearer and the code is somewhat shorter (however I did not test this ... maybe I have bugs)

Reply:
Also maybe taking it a bit far, but having some enum for state saying "unknown, active, inactive" could be even more readable:

```cpp
SmokeCoAlarmServer::ExpressedStateActivation SmokeCoAlarmServer::ComputeActivationState(bool known, bool active)
{
    return known ? (active ? ExpressedStateActivation::kActive : ExpressedStateActivation::kInactive)
                 : ExpressedStateActivation::kInactive;
}

SmokeCoAlarmServer::ExpressedStateActivation SmokeCoAlarmServer::GetStateActive(ExpressedStateEnum e, chip::EndpointId endpointId)
{
    AlarmStateEnum alarmState          = AlarmStateEnum::kNormal;
    EndOfServiceEnum endOfServiceState = EndOfServiceEnum::kNormal;
    bool active                        = false;

    switch (e)
    {
    case ExpressedStateEnum::kSmokeAlarm:
        return ComputeActivationState(GetSmokeState(endpointId, alarmState), alarmState != AlarmStateEnum::kNormal);
    case ExpressedStateEnum::kCOAlarm:
        return ComputeActivationState(GetCOState(endpointId, alarmState), alarmState != AlarmStateEnum::kNormal);
    case ExpressedStateEnum::kBatteryAlert:
        return ComputeActivationState(GetBatteryAlert(endpointId, alarmState), alarmState != AlarmStateEnum::kNormal);
    case ExpressedStateEnum::kInterconnectSmoke:
        return ComputeActivationState(GetInterconnectSmokeAlarm(endpointId, alarmState), alarmState != AlarmStateEnum::kNormal);
    case ExpressedStateEnum::kInterconnectCO:
        return ComputeActivationState(GetInterconnectCOAlarm(endpointId, alarmState), alarmState != AlarmStateEnum::kNormal);
    case ExpressedStateEnum::kEndOfService:
        return ComputeActivationState(GetEndOfServiceAlert(endpointId, endOfServiceState),
                                      endOfServiceState != EndOfServiceEnum::kNormal);
    case ExpressedStateEnum::kTesting:
        return ComputeActivationState(GetTestInProgress(endpointId, active), active);
    case ExpressedStateEnum::kHardwareFault:
        return ComputeActivationState(GetHardwareFaultAlert(endpointId, active), active);
    default:
        return ExpressedStateActivation::kInactive;
    }
}

bool SmokeCoAlarmServer::AutoSetExpressedState(EndpointId endpointId,
                                               std::array<ExpressedStateEnum, kPriorityOrderLength> priorityOrder)
{
    for (ExpressedStateEnum priority : priorityOrder)
    {
        switch (GetStateActive(priority, endpointId))
        {
        case ExpressedStateActivation::kUnknown:
            return false; // internal error
        case ExpressedStateActivation::kActive:
            return SetExpressedState(endpointId, priority);
        case ExpressedStateActivation::kInactive:
            break; // keep searching
        }
    }

    return SetExpressedState(endpointId, ExpressedStateEnum::kNormal);
}
```

Reply:
It's simplified.